ENH: Add zero_phase kwarg to signal.decimate

e-q opened the pull request:
Change the default behavior of ``decimate`` to use ``filtfilt`` instead of
``lfilter``, to eliminate the generally unwanted phase shift between the input
and output data. Changes tests only in that input data is lengthened to satisfy
filtfilt.

Adapted from https://gist.github.com/FRidh/ba60084570d05c91e9bd

Closes #4929

rgommers commented:
Thanks @e-q, looks like a good addition at first sight. 

There's an issue with backwards compat here - numerical results will change when this PR is merged. Default for the new kw should be False to prevent that right? It's a bit difficult to argue that the current behavior is a bug I think. And if it is, then the new kw isn't needed (directly replacing ``lfilter`` with ``filtfilt`` would be better in that case).

rgommers commented:
Test coverage seems to be lacking for ``decimate`` here: there should have been a failing test with the changed default. Could you add a test that checks the numerical values of the result in addition to the shape checks in the current tests?

FRidh commented:
Thanks Erik for opening a PR for this. I agree with Ralf that the default for `zero_phase` should be `False`. We don't want to break anyone's code here. 

I am wondering what a good but simple test would be. Perhaps compare the unwrapped instantaneous phase obtained via the Hilbert transform? Or perhaps easier is to use a cross correlation to check what the shift is for a single sine.

haudren commented:
I have to agree that backwards-compatibility is of greatest importance. However, I feel that typical usage is offline processing of data, and thus the typical use case is `zero_phase is True`, this will need to be explicit in the documentation.

For the tests, if you would like, you can use the snippet I posted in the issue.

e-q commented:
I understand the importance of backwards compatibility, but in reality my perception of this issue is that the use of `lfilter` in this function whatsoever is a bug, rather than something that should be an option. I can't imagine a case where I would accept distorting the phase of a signal when decimating. 

Nevertheless, I'm happy to keep the current behavior as default, but as soon as there is a chance to break backwards compatibility (v1.0?), I would strongly advocate for doing so.

WarrenWeckesser commented on scipy/signal/signaltools.py:
```diff
@@ -2645,11 +2642,16 @@ def decimate(x, q, n=None, ftype='iir', axis=-1):
     axis : int, optional
         The axis along which to decimate.
+    zero_phase : bool
+        Prevent phase shit by filtering with ``fitfilt`` instead of
```
I seriously laughed out load here.  I guess you *really* don't like that distortion, do you? :smile:

FRidh replied:
Ha, I don't think I had that included in my gist! :-)

rgommers replied:
Besides the typo, if you surround function names in the current namespace with single backticks, they get turned into links in the html docs. So goes for ``filtfilt`` here and ``lti`` above.

rgommers replied:
Ah, and didn't even see the worst typo first time around :)

FRidh commented:
@e-q  Actually, there are cases where it does not matter. For example, as acoustician I calculate sound pressure levels with a certain frequency weighting. I apply the weighting with a filter (okay, fair enough, I do not decimate, but just apply the filter), and in that case I have the option between `lfilter` and `filtfilt`. Since a level is calculated as an average over 125 ms I don't care about any phase shift.

I can imagine similar cases when using `decimate`, where after the operation you reduce your amount of samples in a certain way for which the shift does not matter. 

And yes, when we can break backwards compatibility, then let's change the default. I agree on that. It makes more sense to prevent the phase shift, and those few extra cpu cycles shouldn't matter in that case.

rgommers commented:
Just breaking backwards compat in v1.0 is still not the best idea. So here's another proposal:

- make it ``zero_phase=None`` now.
- issue a ``UserWarning`` for ``if zero_phase is None``, saying that the default is now false for bwcompat but will change to True in v0.19.
- open an issue to make the actual change in v0.19 to make sure we don't forget

haudren commented:
:+1: 

I like this proposal, as it would merge the change faster ! (But fix the typos before merging :) )

WarrenWeckesser commented:
If the default filter order for the `lfilter` case is 8, then a reasonable default order for the `filtfilt` case is 4, since that filter is applied to the signal twice, effectively doubling its order.

e-q commented:
Ouch, sorry about all the typos!

I've added a test that compares the expected phase shift of a pure sinusoidal signal with the function output. In practice they only match at a little less than 1%, which I'm not sure how to explain at this point. Nevertheless, this is more than enough to detect whether `zero_phase` is used or not. 

I've added the warning @rgommers suggested, I'm totally onboard with this procedure. 

As to @WarrenWeckesser's comment, your logic is totally sound. However, the current filter design and default order is taken pretty much verbatim from the MATLAB function, where using `filtfilt` is the _only_ option. So, I think the effectively 16th order filter that will be applied to the default `zero_phase=True` use case is still appropriate. 

When/if the default behavior changes, one could imaging doing doubling the order for the `zero_phase=False` case, but I didn't want to get into that kind of thing in this PR.

e-q commented:
I just wanted to check in on this, if anyone has had time to take another look. I'm happy to hear any feedback.

Eric89GXL commented:
Changes look good to me.

FYI after `upfirdn` (#5186) is merged, it should be possible to allow `decimate` to use it instead, which might be faster than using `filtfilt` followed by decimation for some use cases (depending on FIR filter length and downsampling factor). But that's a discussion for later :)

TomDLT commented on scipy/signal/signaltools.py:
```diff
@@ -2660,24 +2662,43 @@ def decimate(x, q, n=None, ftype='iir', axis=-1):
+                      Note: zero_phase will default to True in scipy v0.19.
+                      Until then, decimate defaults to one-way filtering for
+                      backwards compatibility. Ideally, always set this kwarg
+                      excplicitly.
```
typo

e-q commented:
Looks like this didn't make it into v0.17. I'll change the language in the deprecation warning, rebase, etc.

Eric89GXL commented:
@e-q you could now try using `scipy.signal.resample_poly` to see if it's faster than `filtfilt` for some cases. Not sure whether or not it will be. It does zero-phase resampling using polyphase filtering, which skips a bunch of unnecessary multiplies when you know you will upsample or (in this case) downsample afterward.

rgommers commented on scipy/signal/signaltools.py:
```diff
@@ -2680,11 +2681,7 @@ def sosfilt(sos, x, axis=-1, zi=None):
-
-
-def decimate(x, q, n=None, ftype='iir', axis=-1):
+def decimate(x, q, n=None, ftype='iir', axis=-1, zero_phase=False):
```
This should be ``zero_phase=None``

rgommers commented on scipy/signal/signaltools.py:
```diff
@@ -2696,13 +2693,18 @@ def decimate(x, q, n=None, ftype='iir', axis=-1):
+        of an `lti` object, uses that object to filter before downsampling.
     axis : int, optional
         The axis along which to decimate.
+    zero_phase : bool
```
``, optional``

rgommers commented on scipy/signal/signaltools.py:
```diff
@@ -2696,13 +2693,18 @@ def decimate(x, q, n=None, ftype='iir', axis=-1):
     axis : int, optional
         The axis along which to decimate.
+    zero_phase : bool
+        Prevent phase shift by filtering with `filtfilt` instead of `lfilter`.
```
Add here what ``None`` does, and that it defaults to ``False`` for backwards compat. And that ``True`` is recommended.

rgommers commented on scipy/signal/signaltools.py:
```diff
@@ -2713,24 +2715,43 @@ def decimate(x, q, n=None, ftype='iir', axis=-1):
+                      Until then, decimate defaults to one-way filtering for
+                      backwards compatibility. Ideally, always set this kwarg
+                      explicitly.
+                      """)
```
Can you make this a ``FutureWarning``? 

And better make it simply "in a future release" rather than v0.20. Because timing of making changes varies, and because I'd like to never get to v0.20 (but 1.0 before that).

rgommers commented:
@e-q sorry that this missed v0.17. I made a few minor comments, and the PR needs a rebase. Then I think it's good to merge. Trying ``signal.resample_poly`` as @Eric89GXL suggests could be a nice further improvement, but that can also be done after merging.

e-q commented:
Thanks for the feedback, I'll have it updated soon. 

I do have a question about scope, though. As I noted in recently in #4929, `filtfilt` maybe isn't the best option for the FIR path of `decimate`, due to the memory use. There is another method for FIR decimation that involves accounting for the linear group delay and setting up the filter's initial conditions to match the initial points. I haven't yet written code for this, so should I restrict this PR to introducing `filtfilt` to the IIR path and make another PR for the FIR path, or delay this PR a bit until both are ready?

Eric89GXL commented:
@e-q it sounds like the zero-phase kwarg here and the proposed separate FIR path are separable (enough) issues, so Id' recommend 2 PRs to reduce review burden.

rgommers commented:
I think I agree - just noting in the ``zero_phase`` description that it applies only to IIR should make this mergeable and not create issues when later adding the improvement for FIR.

e-q commented:
Ok, I've made changes to restrict this to only take effect when `ftype='iir'`, added the appropriate language, and squashed it all together.

Eric89GXL commented on scipy/signal/signaltools.py:
```diff
@@ -2826,13 +2823,21 @@ def decimate(x, q, n=None, ftype='iir', axis=-1):
+        Prevent phase shift by filtering with `filtfilt` instead of `lfilter`
+        when `ftype` is not 'fir'.  A value of `True` is recommended, since a
+        phase shift is generally not desired. Using `None` defaults to `False`
+        for backwards compatibility.
```
add `.. versionadded: 0.18` directive

e-q replied:
Done, thanks!

Eric89GXL commented on scipy/signal/signaltools.py:
```diff
@@ -2843,24 +2848,49 @@ def decimate(x, q, n=None, ftype='iir', axis=-1):
+                      explicitly.""", FutureWarning)
+        zero_phase = False
+
+    if zero_phase is True and ftype == 'fir':
```
I slightly prefer `if zero_phase` to `if zero_phase is True` because it allows for reasonable/expected behavior for e.g. `zero_phase=1` and `zero_phase=0`.

e-q replied:
Fair enough, changed

Eric89GXL commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -1380,17 +1380,56 @@ def test_filtfilt_gust():
+        # Decimate, see if response is in phase with original
+        xdec = signal.decimate(d.real, q, n, system, zero_phase=True)
+        phase = np.mean(np.angle(d[::q].conj()*xdec))
+        assert_almost_equal(phase, 0, decimal=3)
```
I recently added a few test cases to `test_resample`. Maybe it's worth unifying to test on some of the same signals, since decimation can be thought of as a special case of resampling?

e-q commented:
Based on @Eric89GXL's comment, I reworked the phase shift tests to be more like the ones he wrote for `resample`, so now much more than one sinusoid at one downsampling factor is tested. 

While doing this, I noticed that the default downsampling filters do not meet the residual magnitude requirements set in `test_resample` for frequencies above the new nyquist. Maybe we'll revisit them for v1.0, when I assume we can break backwards compatibility for this kind of thing?

Eric89GXL commented:
There will be tradeoffs between filter steepness (and thus attenuation),
side band attenuation, temporal ringing, etc. based on the filter that is
used, so it's possible what's done in decimate is already good enough for
some applications, and the tolerances I put in for resample_poly aren't
ideal for those cases.

It's probably just worth noting the differences in code and preferably in
docstring somewhere, especially if you can identify how they are different
(e.g. filter length or construction parameters for FIR). Unifying the
treatments should probably be a separate PR because we'll probably need to
discuss (and at least for me, read about) the right thing to do.

e-q commented:
Hmm, it's only failing on Python 3.4...

Eric89GXL commented:
(Without looking at the code) that could indicate a float vs integer
division problem. If you can't track it down let me know and I'll look once
I'm back at a proper computer.

haudren commented:
Apparently the error comes from a pi that should be a -pi. I am on my phone now but I think this comes from some kind of rounding error and the resulting call to numpy.angle gives opposite results.

e-q commented:
I adjusted the tests to try and avoid -pi and pi looking so far apart. Locally passes for me on Python 2.7 and 3.4.

Eric89GXL commented on scipy/signal/signaltools.py:
```diff
@@ -2843,24 +2850,52 @@ def decimate(x, q, n=None, ftype='iir', axis=-1):
+    if zero_phase is None:
+        warnings.warn("""
+                      Note: Decimate's zero_phase keyword argument will default
+                      to True in a future release.  Until then, decimate
```
won't using `"""` here make it so your warning has line breaks, plus a large indent?

e-q replied:
Yes. Oops.

Eric89GXL commented:
Changes LGTM. Can you see any good way to unify the tests here with those for `resample` and `resample_poly`?

e-q commented:
> Can you see any good way to unify the tests here with those for `resample` and `resample_poly`?

I adopted the structure of the `resample` tests, but at this point I think they need to stay a little different, since we need to test the phase shifting behavior for `decimate` for the time being.

e-q commented:
Had to rebase. Once this is in, I'll be starting a PR for fixing up the FIR path.

Eric89GXL commented:
Thanks. Anyone else want to take a look? @rgommers you had a look before if you want to look again. I'll merge in a couple days if nobody else chimes in.

Eric89GXL commented:
Thanks @e-q
